Remove Shell-Specific Logic (#168)
buildpacks/rfcs: text/0000-remove-shell-processes.md

# What it is
[what-it-is]: #what-it-is

## Process Schema
The existing schema for the processes table in `launch.toml` is
```
# Old Schema

[[processes]]
type = "required"
command = "required"
args = ["optional"]
direct = false
default = false
```

The new proposed schema is:
```
# New Schema

[[processes]]
type = "required"
command = ["required"]
args = ["optional"]
default = false
```
The following changes have been made:
- **`direct` has been removed** - all processes are executed directly. If `bash` or `cmd.exe` is required it should be included in `command`. No surprises.
- **`command` is now an array** - Arguments in `command` *will not* be overwritten if a user provides additional arguments at runtime. `args` are default arguments that *will* be overwritten if a user provides additional arguments at runtime. The makes `command` analogous to `Entrypoint` in the OCI spec and `command` in a Kubernetes PodSpec. `args` is analogous to `cmd` and `args` in docker and Kubernetes respectively.

## Using Environment Variables in a Process

> One upside to our previous execution strategy was that it enable users to include environment variable references in arguments that were later evaluated in the container. To preserve this feature we can instead adopt a variabtion on the Kubernetes strategy for [environment variables interpolation](https://kubernetes.io/docs/tasks/inject-data-application/define-environment-variable-container/#using-environment-variables-inside-of-your-config). If a buildpack or user includes `${<env>}` in the `command` or `args` and `{env}` is the name of an environment variable set in the launch environment, the launcher will replace this string with the value of the environment variable after apply buildpack-provided env modifications and before launching the process.

Reviewer: ```suggestion
One upside to our previous execution strategy was that it enable users to include environment variable references in arguments that were later evaluated in the container. To preserve this feature we can instead adopt a variation on the Kubernetes strategy for [environment variables interpolation](https://kubernetes.io/docs/tasks/inject-data-application/define-environment-variable-container/#using-environment-variables-inside-of-your-config). If a buildpack or user includes `${<env>}` in the `command` or `args` and `{env}` is the name of an environment variable set in the launch environment, the launcher will replace this string with the value of the environment variable after apply buildpack-provided env modifications and before launching the process.
```

> One upside to our previous execution strategy was that it enable users to include environment variable references in arguments that were later evaluated in the container. To preserve this feature we can instead adopt a variabtion on the Kubernetes strategy for [environment variables interpolation](https://kubernetes.io/docs/tasks/inject-data-application/define-environment-variable-container/#using-environment-variables-inside-of-your-config). If a buildpack or user includes `${<env>}` in the `command` or `args` and `{env}` is the name of an environment variable set in the launch environment, the launcher will replace this string with the value of the environment variable after apply buildpack-provided env modifications and before launching the process.

Reviewer 2: Is there a risk that in doing this we might somehow expose secrets? For example if the process type is:

```
./bin/server --key $MY_SECRET
```

If we substitute `MY_SECRET` in the process could it end up in the logs or somewhere that it might not when Bash substitutes it? 

Are they known libraries/mechanisms we can use to make sure we're on the right-side of the law here?

Author: I think we are right that there is some cause for concern here. I think we should change this RFC and switch the interpolation notation to something like `$((MY_SECRET))` so we never interpolate environment variables that are intended for Bash to interpolate in a command like `["bash", "-c", "./bin/server --key ${MY_SECRET}"]. It's a little sad that we need yet another notation but it seems good to remove any ambiguity with regards to what gets interpolated by k8s, vs the launcher, vs Bash (if it is explicitly included in a command).

In this example are you imagining that it is the user you provides `$MY_SECRET` dynamically at runtime? I would argue that this is not a safe thing to do with secrets because they end up in the container config. Secrets should be provided via a safe mechanism like bindings and buildpacks should be responsibility for protecting the secret, which hopefully will not require passing it directly on the command line (even as a reference for Bash to interpolate) in the vast majority of use cases.

Reviewer 3: `$(ENV)` should work? (the k8s convention). It only works in k8s for env vars directly defined  in the spec, the rest are left untouched.

Author: updated to use k8s style notation in https://github.com/buildpacks/rfcs/pull/168/commits/827cdd5782eceed7ba367941af03dd8d519bb3d3

# How it Works
[how-it-works]: #how-it-works

## Buildpack-provided process types
### Example 1 - A Shell Process
The Paketo .Net Execute Buildpack may generates shell processes similar to the following:
```
[[processes]]
type = "web"
command = "dotnet my-app.dll --url http://0.0.0.0:${PORT:-8080}"
direct = false
```
NOTE: the buildpack API used by this buildpack (`0.5`) predates the introduction of `default`.

Using the new API this process could look like:
```
[[processes]]
type = "web"

> command = ["dotnet", "my-app.dll", "--urls", "http://0.0.0.0:${PORT}"] # the default value of PORT would need to be provided in a layer

Reviewer: ```suggestion
command = ["dotnet", "my-app.dll", "--url", "http://0.0.0.0:${PORT}"] # the default value of PORT would need to be provided in a layer
```

Author: I'll admit to knowing precious little about the `dotnet` CLI, but I copied this command directly from the buildpack. And it seems to agree with [this blog post](https://andrewlock.net/5-ways-to-set-the-urls-for-an-aspnetcore-app/#command-line-arguments) I just found.

Author: Sorry, I think I found the source of confusion - https://github.com/buildpacks/rfcs/pull/168/commits/c97e7f22e989bb6aa45570d8f7da1ca86631ffa8

Author: I'm gonna assume this can be resovled

default = true
```
Things to note:
* In the above example I have eliminated the dependency on Bash instead of explicitly adding it to the command, because it is likely unnecessary.

> * If the buildpack authors believed that `--urls` should be overridable they could set move the last two arguments from `command` to `args`.

Reviewer: ```suggestion
* If the buildpack authors believed that `--url` should be overridable they could set move the last two arguments from `command` to `args`.
```

Author: https://github.com/buildpacks/rfcs/pull/168#discussion_r659030633

Author: https://github.com/buildpacks/rfcs/pull/168#discussion_r661799403

### Example 1 - A Direct process
The follow direct commands:
```
docker run --entrypoint launcher <image> -- env
docker run --entrypoint launcher <image> -- echo hello '$WORLD' 
```
will become the following, using the new platform API
```
docker run --entrypoint launcher <image> env
docker run --entrypoint launcher <image> echo hello '${WORLD}'
```
Previously, in the second command in this example `${WORLD}` would not have been interpolated because this is a direct process; instead the output would include the literal string `$WORLD`. With the changes proposed, `${WORLD}` will now be evaluated, even though the process is direct.

### Example 2 - A Shell Process
The follow custom shell command:
```
docker run --entrypoint launcher <image> echo hello '${WORLD}'
docker run --entrypoint launcher <image> echo hello '${WORLD:-world}'
```
will become the following, using the new platform API
```
docker run --entrypoint launcher <image> echo hello '${WORLD}'
docker run --entrypoint launcher <image> bash -c 'echo hello "${WORLD:-world}"'
```

> The first command in this example did not need to change to behave as expected with the new API. Previously it was necessary to use a shell process in order to evaluate `${WORLD}`. Now, the shell is unescessary. However, if the user wants to use more features of Bash interpolation they will need to explicitly invoke a shell, as shown in the second command in this example.

Reviewer 2: ```suggestion
The first command in this example did not need to change to behave as expected with the new API. Previously it was necessary to use a shell process in order to evaluate `${WORLD}`. Now, the shell is unnescessary. However, if the user wants to use more features of Bash interpolation they will need to explicitly invoke a shell, as shown in the second command in this example.
```

Reviewer 2: I'm not sure I can complete sense of what "more features of Bash interpolation" includes. In particular, does this include things like `;`, `&`, and `&&` in the command? Those are the most common patterns I can think of (aside from env vars). What about different methods of quoting (`"` versus `'`)?

Author: To avoid this whole question I think we should switch from `${NAME}` to `$((NAME))` for our env var notation. Therefore nobody is tempted to try things like `${NAME:-default-name}`. We should keep it as simple as possible and avoid surprises.

### Example 3 - A Script Process
The follow custom script command:
```

> docker run --entrypoint launcher <image> 'for opt in $JAVA_OPTS; do echo $opt; done'

Reviewer 2: One of the things I worry about in this example is how it might unexpectedly fail. I expect that most users will do the following:

1. upgrade
2. try the exact same command they used before (i.e. no `bash -c`)
3. watch it fail
4. realize they need to add `bash -c`

In step 3, is it possible that something very bad happens if their process kinda works but not as they expected?

Author: I agree that they might upgrade and fail but I don't think there is any danger of the process even kinda working. In this case the launcher would be trying to directly invoke an executable called `'for opt in $JAVA_OPTS; do echo $opt; done'`

```
will become the following, using the new platform API
```
docker run --entrypoint launcher <image> bash -c 'for opt in $JAVA_OPTS; do echo $opt; done'
```

## What About Profile Scripts?

### `<layer>/profile.d/*`
When a buildpack upgrades to the new buildpack API, buildpack-provided profile scripts will no longer be supported. Instead buildpacks can use `exec.d`. Most existing profiles scripts can be easily converted.

**Example 1: `jkutner/sshd` **

The following profile script example was taken from https://github.com/jkutner/sshd-buildpack/blob/master/sbin/sshd.sh
```
#!/usr/bin/env bash

if [[ "${SSH_DISABLED:-}" != "true" ]]; then
  ssh_layer=$(realpath $(dirname ${BASH_SOURCE[0]})/..)
  ssh_dir=$(realpath $HOME/.ssh)

  mkdir -p $ssh_dir
  cat $ssh_layer/id_rsa.pub >> $ssh_dir/authorized_keys

  cat << EOF >> $ssh_dir/sshd_config
HostKey $ssh_layer/id_rsa
AuthorizedKeysFile $ssh_dir/authorized_keys
EOF

  chmod 600 /workspace/.ssh/*

  ssh_port=${SSH_PORT:-"2222"}
  echo "at=sshd state=starting user=$(whoami) port=${ssh_port}"
  /usr/sbin/sshd -f $ssh_dir/sshd_config -o "Port ${ssh_port}"
fi
```

> This script does not require any changes and would work as an `exec.d` or `profile.d`.

Reviewer 2: I'd still need to move the script in the `exec.d` though right? are there any restrictions like filename or permissions that might change? (I'm sure these are all fine but would like to capture)

Author: Good point. We should capture two things:
1. The script must move to `<layer>/exec.d/`
2. The runtime user must have permission to execute this file (e.g. executable bit should be set).

Author: Added in https://github.com/buildpacks/rfcs/pull/168/commits/b49861f13dd11b080099b0da7ac8b49e9611f113

**Example 2: paketo-buildpacks/node-engine **

The following profile script example was taken from https://github.com/paketo-buildpacks/node-engine/blob/main/environment.go
```
if [[ -z "$MEMORY_AVAILABLE" ]]; then
		memory_in_bytes="$(cat /sys/fs/cgroup/memory/memory.limit_in_bytes)"
		MEMORY_AVAILABLE="$(( $memory_in_bytes / ( 1024 * 1024 ) ))"
fi
export MEMORY_AVAILABLE
```

The following revised version of this script would implement the `exec.d` interface
```
if [[ -z "$MEMORY_AVAILABLE" ]]; then
		memory_in_bytes="$(cat /sys/fs/cgroup/memory/memory.limit_in_bytes)"
		MEMORY_AVAILABLE="$(( $memory_in_bytes / ( 1024 * 1024 ) ))"
fi

cat << EOF >&3
MEMORY_AVAILABLE = "${MEMORY_AVAILABLE}" 
EOF
```

### `<app>/.profile`

> When a platform upgrades to the new platform API, user-provided profile scripts will no longer be supported. A buildpack could provide an identical or similar interface by detecting a `.profile` file in the app dir, wrapping it so that it implements the `exec.d` interface and add creating an executable with that layer. A platform that wishes to always provide that functionality could add a buildpack like this to every build. By making dynamic runtime environment modification a buildpack concern and not duplicating similar behavior in the platform API, we reduce the surface area and complexity of the platform API and thus improve normalization of the spec.

Reviewer 2: Is this true even when the use `bash -c` in their command?

Author: Yes I believe this is still true. `exec.d` helpers currently run for both direct and shell processes.

Reviewer 2: No, i mean if they use `bash -c` will `.profile` still be not be run? I feel like that might be unexpected.

Author: Yes, `.profile` will still not be run even when the command includes `bash -c`. Part of the goal here is to avoid doing anything complicated or magic. In order to run `.profile`s we would need to change the provided command from something like `['bash', '-c', 'echo hi']` to `['bash', '-c', 'source <app>/.profile && echo hi']`. Modifying the provided command if and only if it starts with `bash -c` feel like too much instrusive magic to me.

```

and the launcher should Bash evaluate each entry in `command` to avoid breaking changes.

However, any additional user-provided args should NOT be Bash evaluated to reduce the amount of complexity exposed to end users.

### profiles
Profiles contributed by older buildpacks will still be evaluated when executing shell process types. But user provided profiles will not be evaluated when using the new platform API, even when a shell process is executed. Again, this is done to prevent differences between the buildpack APIs from leaking into the user interface. Users will only need think about differences between platform APIs.

# Drawbacks
[drawbacks]: #drawbacks

In exchange for a reduction in complexity and cognitive overhead buildpack-authors and end users lose certain conveniences like the more intuitive profile interface. This drawback could be remediated by tooling for buildpack authors (to make creation of `exec.d` helpers easy) or by buildpacks (to support `.profile`). However, in the `.profile` case, consistency across ecosystems will be a matter of convention (like Procfile) rather than a guarantee.

# Alternatives
[alternatives]: #alternatives

## `$(<env>)` syntax

> We could use `$(<env>)` for environment variable replacements instead of `${<env>}`. urrently, Kubernetes only replaces `$(FOO)` if `FOO` matches the name of an enivornment variable defined in the `env` section of the PodSpec, otherwise `$(FOO)` is passed through literally. The would allow thelauncher to pick up where kubernetes leaves off and evaluate an remaining references.

Reviewer 4: ```suggestion
We could use `$(<env>)` for environment variable replacements instead of `${<env>}`. Currently, Kubernetes only replaces `$(FOO)` if `FOO` matches the name of an environment variable defined in the `env` section of the PodSpec, otherwise `$(FOO)` is passed through literally. The would allow the launcher to pick up where kubernetes leaves off and evaluate any remaining references.
```

pros:
* Users can continue to use a syntax they are already familiar with
cons:
* User provided variables may be evaluated too soon, before buildpack modifications are applied.

## Lifecycle support for `<app>/.exec`

> When we remove support for `<app>/.profile` we could add support for `<app>/.exec` or similar where `<app>/.exec` must implement the `exec.d` interface.

Reviewer 2: I'd like to bring this into scope

Author: After thinking long and hard about it I am very opposed to bringing this into scope. I don't think contracts between the app source and the lifecycle are supportable. The platform API should be a contract between a platform and the lifecycle, to enable predictable behavior and a clean upgrade story. Otherwise the same app will behave different on different platforms depending on the platform API in use. I realize the user must have some awareness of the "launch" api to understand argument handling for the resulting image but at least they won't need to change their source code to run on different platforms. This RFC is trying hard to shrink the surface area of platform API, that the app developer needs to be aware of.

If we wanted to respect config like this in the app dir, a buildpack should provide that functionality. App developers can included the exact buildpack they want to get the exact outcome they want without having to keep track of the versions of the platform API currently implemented by the platforms they use (which is entirely outside their control).

Reviewer: @ekcasey isn't this already true by us removing support fro `<app>/.profile` since this changes the contract? I get you're trying to simplify/minimize the scope here, but bringing this into scope or not is still changing that contract.

Author: While you are right that removing `<app>/.profile` modifies this contract, it shrinks it in a way that should reduce future occurrences of this problem. Adding `<app>/.exec` modified this contract without helping to mitigate this problem in the future.

Reviewer 2: I've created a an RFC that proposes a new mechanism to replace the `.profile` behavior (thus moving the contract to be between a builder<>app). https://github.com/buildpacks/rfcs/pull/175

pros:
* Users may still dynamically modify the runtime environment without requiring a specific buildpack
cons:
* The exec.d interface must be duplicated in both the buildpack and platform API
* If we modify the interface the same app might behave differently or fail to run on specific platforms depending on which version of the platform API they are using

> * Users must directly implement our less-than-perfectly-intuitive exec.d interface instead of whatever better UX buildpack authors invent.

Author: > Often, those modifications aren't even buildpack specific, but app specific

Not sure I follow why this is relevant.

What I am trying to say is that a single buildpack could continue to provide support for .profile scripts without forcing users to migrate to the exec.d interface, if that was desirable.

* launcher behavior that fails on certain stacks
* undesirable coupling between the lifecycle and specific shells

## API compat strategy - Explicit Bash Evaluation
When adding legacy shell processes to `metadata.toml` we could replace the "direct=false" command with the literal command that will be evaluated.
For example the following script process in `launch.toml`
```
[[process]]
type = "hi"
command = "echo hello "${WORLD:-world}"
direct = false
```
Could become the following in `metadata.toml`
```
[[process]]
type = "hi"
command = ["bash", "-c", 'echo hello "${WORLD:-world}"']
direct = true # we could even potentially remove this from the metadata.toml schema entirely
```

However, things get very complicated if we explicitly source profiles or shell evaluated args in `command`. The complexity of the resulting commands probably makes this strategy untenable.

# Prior Art
[prior-art]: #prior-art

The Paketo Java buildpacks have already converted an extensive collection of profile script to exec.d binaries and converted all processes to direct processes in order to support minimal stacks like `io.paketo.stacks.tiny` and to provide users with more intuitive argument handling.

> The Procfile interface is supported by Paketo, Heroku, and Google buildpacks, demonstrating that it is possible to have a consistent interface across buildpack ecosystems without building direct support for that interface in the lifecycle. 

Reviewer 2: I'm not sure this is winning example. There are 3 different impls of something that should (IMO) be a native feature. The fact that these haven't converged on a single Procfile buildpacks is worrisome and not a pattern I'd like to repeat.

Author: I love to see an RFC to provide utility buildpacks as part of the CNB project, that implement extension specs or common conventions.

Reviewer 2: https://github.com/buildpacks/rfcs/pull/175

### Example 1 - A Direct process
The follow direct commands:
```
docker run --entrypoint launcher <image> -- env
docker run --entrypoint launcher <image> -- echo hello '$WORLD' 
```
will become the following, using the new platform API
```
docker run --entrypoint launcher <image> env
docker run --entrypoint launcher <image> echo hello '${WORLD}'
```
Previously, in the second command in this example `${WORLD}` would not have been interpolated because this is a direct process; instead the output would include the literal string `$WORLD`. With the changes proposed, `${WORLD}` will now be evaluated, even though the process is direct.

### Example 2 - A Shell Process
The follow custom shell command:
```
docker run --entrypoint launcher <image> echo hello '${WORLD}'
docker run --entrypoint launcher <image> echo hello '${WORLD:-world}'
```
will become the following, using the new platform API
```
docker run --entrypoint launcher <image> echo hello '${WORLD}'

> docker run --entrypoint launcher <image> bash -c 'echo hello "${WORLD:-world}"'

Reviewer 3: Could we support these use cases by default? GIven that most envsubst libs already support it - 

https://github.com/a8m/envsubst
https://github.com/drone/envsubst

This would also be useful as it would end up in the metadata and users could see default values for these variables as set by processes.

Author: I _just_ decided to take this in a different direction - https://github.com/buildpacks/rfcs/pull/168/commits/827cdd5782eceed7ba367941af03dd8d519bb3d3

Author: I found some of the comment from the original k8s implementation of this illuminating - https://github.com/kubernetes/kubernetes/pull/7936#issuecomment-100950516

Reviewer 3: Was going through the same thing.

On a separate note - I guess it should be a trivial change in the above libs to add support for a different marker instead of `{`. WIll see if they are open to it. OTOH I guess we could always start simple and add it later if needed. I would've like a way to specify a default value at the very least though.

Reviewer 3: FWIW - just needs a change at https://github.com/drone/envsubst/blob/7bf45dbf5372da60a12eab36b9c8a63c99b92ea5/parse/scan.go#L174

Author: @samj1912 can this comment be resolved? or do you still want to advocate for more env substitution functionality, including defaults?

Personally I think we should go with the simple version for now.

- The difference between a direct and shell process
- The nuances of argument handling in direct vs. Bash vs. Cmd cases
- That buildpack-provided profile scripts will not apply to direct processes
- That a user-provided `.profile` script will not apply to a direct processes

 Removing this special behavior and requiring buildpacks to explicitly include any required shell in the command itself creates a simpler, more comprehensible model for buildpack authors and end users alike.

# What it is
[what-it-is]: #what-it-is

## Process Schema
The existing schema for the processes table in `launch.toml` is
```
# Old Schema

[[processes]]
type = "required"
command = "required"
args = ["optional"]
direct = false
default = false
```

The new proposed schema is:
```
# New Schema

[[processes]]
type = "required"
command = ["required"]
args = ["optional"]
default = false
```
The following changes have been made:

> - **`direct` has been removed** - all processes are executed directly. If `bash` or `cmd.exe` is required it should be included in `command`. No surprises.

Reviewer 5: I think there could be some nuance here that we need to document when switching from a shell process to direct regarding signals. Many platforms support the concept of a graceful shutdown, which is a SIGTERM followed by a pause, followed by SIGKILL on your process.

When you run as a shell process, pid 1 is your shell which generally traps and handles signals as you'd expect. When you switch to running your process as direct, it becomes pid 1 and it may or may not handle all the signals correctly (you can generally tell if it's not handling them correctly by running a container and trying to interrupt it with ctrl+c, if that works it's handling signals correctly). 

Anyway, I think we need to call out this consideration so that when buildpack authors are migrating their buildpacks they think about this nuance and ensure it's handled correctly.

Author: After some reading I suspect the default signal handling behavior is actually better for a direct process than a shell process. Bash does not forward the SIGTERM to the child process by default and therefore there is no opportunity for a truly graceful shutdown even if the server process has signal handling logic. Therefore a buildpack author that is not thinking about this explicitly might prevent the appropriate handling of a sigterm if they default to creating a shell process.

https://veithen.io/2014/11/16/sigterm-propagation.html

cc @hone, @dmikusa-pivotal

Reviewer 6: > therefore there is no opportunity for a truly graceful shutdown even if the server process has signal handling logic. 

Surprisingly, it's even worse than this. A simple bash invocation (e.g., `bash -c "tail -f /dev/null"`) is treated differently by bash than a complex one (e.g., `bash -c "true && tail -f /dev/null"`). In the former case, bash `exec`s itself away, resulting in a single process. So from the user's perspective, whether this works or not is quite arbitrary.

Reviewer 5: Hmm, why isn't anything ever easy?  Lol.

The behavior I'd observed is with apps that don't handle signals at all. If you run them as PID1, you can't terminate the container. It's SIGKILL or the app keeps running. However, if you wrap that in `bash -c` it'll terminate as expected. I guess I didn't dig deep enough to check how Bash was in turn stopping that app, based on what's linked here it is just killing it as well and not propagating signals.

Regardless, there seems to be enough complexity here that I still think this warrants a mention in this RFC document. It's nuanced behavior that could easily be missed by a buildpack author, if we call attention to it, that gives authors a chance to investigate what it means to them. 

Even if it's as simple as "When running a process as direct, your process will always run as PID1. Buildpack authors, please be aware this may change the behavior of your application at shutdown and you should review accordingly.", I think that give's authors a heads up.

Reviewer 5: One more "fun" point on this subject. If you use `sh -c` instead of `bash -c` you get more different behavior because at least on Ubuntu, `/bin/sh` is not `/bin/bash`, it's `/bin/dash`. I was running a process this morning with `sh -c` and it completely ignored SIGTERM.

Author: @sclevine Are you aware of any documentation regarding when bash does or doesn't exec the child process. When I tried out you example commands I observed Bash replacing itself in memory, even for the more complex command:
```
docker run --rm --name test --entrypoint bash ubuntu:bionic -c 'true && tail -f /dev/null'
```
```
docker exec test ps -ef
UID          PID    PPID  C STIME TTY          TIME CMD
root           1       0  0 14:41 ?        00:00:00 tail -f /dev/null
root           7       0  0 14:41 ?        00:00:00 ps -ef
```

Author: Okay, if I am understanding correctly, I think there are relevant two cases here:
1. the command is a complicated Bash invocation
2. the command is a simple Bash invocation

My understanding of the behavior (based on @sclevine 's description and my experimental verification):

**case 1**: A complicated command cannot be executed directly (`true && tail -f /dev/null` is not a valid direct process) and therefore `bash -c` must be explicitly included in the process. Therefore, wrt signal handling, **nothing changes**.
**case 2**: Bash is already `exec`ing itself away leaving the process as PID1. If this is converted to a direct process the same process will be PID1. Therefore, wrt signal handling, **nothing changes**.

**Summary**: With regards to signal handling nothing changes. App developers and buildpacks authors do not need to worry about changes to signal handling.

Author: @dmikusa-pivotal & @hone do you feel that this comment can be resolved given the explanation above describing how this does not affect end users or buildpack authors?

Reviewer 5: I think this discussion has been good around signals. It seems unlikely that it's going to cause issues, so I think we're good there.

Maybe one final suggestion with this sentence...

>If `bash` or `cmd.exe` is required it should be included in `command`.

I think "required" might be nebulous and assumes you know what is required. It's easy to overlook restrictions of running as direct/PID1. Perhaps we could have something like:

>If currently using `direct = false`, the recommended migration path for buildpack authors is to include `bash` or `cmd.exe`  in `command`. This will result in the same behavior as `direct = false`. Before switching to `direct = true`, it is recommended you confirm environment variable interpolation is not required and that your process can function as PID1.

I guess it doesn't necessarily have to be in this line, but I think we should offer some migration advice and list the concerns an author should review before switching to `direct = true`.

Author: @dmikusa-pivotal please see https://github.com/buildpacks/rfcs/pull/168/commits/14066e50a3ee999fe0ff5249362e43c52169cac5

Reviewer 5: Looks great, thanks!

   2. Remove an unnecessary dependency on  `bash` or `cmd` by:
        * Changing any environment variable references to use the `$(<env)` syntax.
        * Ensure that your process functions properly as PID1.

## Platform Maintainers

Before upgrading to the new API platform, platforms that wish to support the `<app>/.profile` or `<app>/.profile.bat` should ensure that all builds contain a buildpack that provides support for this feature, whenever the stack provides the requisite shell.

# Drawbacks
[drawbacks]: #drawbacks

In exchange for a reduction in complexity and cognitive overhead buildpack-authors and end users lose certain conveniences like the more intuitive profile interface. This drawback could be remediated by tooling for buildpack authors (to make creation of `exec.d` helpers easy) or by buildpacks (to support `.profile`). However, in the `.profile` case, consistency across ecosystems will be a matter of convention (like Procfile) rather than a guarantee.

# Alternatives
[alternatives]: #alternatives

## `$((<env>))` syntax

We could use our own explicit syntax like `$((<env>))` or environment variable replacements instead of `$(<env>)`.

> pros:

Reviewer 4: The pros/cons look funny in the PR markdown preview. I think you may need an additional line break between lists

# How it Works
[how-it-works]: #how-it-works

## Buildpack-provided process types
### Example 1 - A Shell Process
The Paketo .Net Execute Buildpack may generates shell processes similar to the following:
```
[[processes]]
type = "web"
command = "dotnet my-app.dll --urls http://0.0.0.0:${PORT:-8080}"
direct = false
```
NOTE: the buildpack API used by this buildpack (`0.5`) predates the introduction of `default`.

Using the new API this process could look like:
```
[[processes]]
type = "web"

> command = ["dotnet", "my-app.dll", "--urls", "http://0.0.0.0:${PORT}"] # the default value of PORT would need to be provided in a layer

Reviewer 7: ```suggestion
command = ["dotnet", "my-app.dll", "--urls", "http://0.0.0.0:$(PORT)"] # the default value of PORT would need to be provided in a layer
```

Should this follow the k8s interpolation format since this is no longer a shell process?

Reviewer: @ekcasey please take a look.
